// @grafana/schema/dist/esm/raw/dashboard/x/dashboard_types.gen.js
const defaultAnnotationTarget = {
  tags: []
};
const defaultAnnotationQuery = {
  builtIn: 0,
  enable: true,
  hide: false,
  showIn: 0,
  type: "dashboard"
};
const defaultDashboardLink = {
  asDropdown: false,
  includeVars: false,
  keepTime: false,
  tags: [],
  targetBlank: false
};
const defaultGridPos = {
  h: 9,
  w: 12,
  x: 0,
  y: 0
};
const defaultThresholdsConfig = {
  steps: []
};
const defaultDashboardCursorSync = 0 /* Off */;
const defaultPanel = {
  links: [],
  repeatDirection: "h",
  tags: [],
  targets: [],
  thresholds: [],
  timeRegions: [],
  transformations: [],
  transparent: false
};
const defaultFieldConfigSource = {
  overrides: []
};
const defaultMatcherConfig = {
  id: ""
};
const defaultFieldConfig = {
  links: [],
  mappings: []
};
const defaultRowPanel = {
  collapsed: false,
  panels: []
};
const defaultDashboard = {
  editable: true,
  graphTooltip: 0 /* Off */,
  links: [],
  panels: [],
  schemaVersion: 36,
  style: "dark",
  tags: [],
  timezone: "browser"
};

export { defaultAnnotationQuery, defaultAnnotationTarget, defaultDashboard, defaultDashboardCursorSync, defaultDashboardLink, defaultFieldConfig, defaultFieldConfigSource, defaultGridPos, defaultMatcherConfig, defaultPanel, defaultRowPanel, defaultThresholdsConfig };
