// @grafana/schema/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var AxisPlacement = /* @__PURE__ */ ((AxisPlacement2) => {
  AxisPlacement2["Auto"] = "auto";
  AxisPlacement2["Bottom"] = "bottom";
  AxisPlacement2["Hidden"] = "hidden";
  AxisPlacement2["Left"] = "left";
  AxisPlacement2["Right"] = "right";
  AxisPlacement2["Top"] = "top";
  return AxisPlacement2;
})(AxisPlacement || {});
var AxisColorMode = /* @__PURE__ */ ((AxisColorMode2) => {
  AxisColorMode2["Series"] = "series";
  AxisColorMode2["Text"] = "text";
  return AxisColorMode2;
})(AxisColorMode || {});
var VisibilityMode = /* @__PURE__ */ ((VisibilityMode2) => {
  VisibilityMode2["Always"] = "always";
  VisibilityMode2["Auto"] = "auto";
  VisibilityMode2["Never"] = "never";
  return VisibilityMode2;
})(VisibilityMode || {});
var GraphDrawStyle = /* @__PURE__ */ ((GraphDrawStyle2) => {
  GraphDrawStyle2["Bars"] = "bars";
  GraphDrawStyle2["Line"] = "line";
  GraphDrawStyle2["Points"] = "points";
  return GraphDrawStyle2;
})(GraphDrawStyle || {});
var GraphTransform = /* @__PURE__ */ ((GraphTransform2) => {
  GraphTransform2["Constant"] = "constant";
  GraphTransform2["NegativeY"] = "negative-Y";
  return GraphTransform2;
})(GraphTransform || {});
var LineInterpolation = /* @__PURE__ */ ((LineInterpolation2) => {
  LineInterpolation2["Linear"] = "linear";
  LineInterpolation2["Smooth"] = "smooth";
  LineInterpolation2["StepAfter"] = "stepAfter";
  LineInterpolation2["StepBefore"] = "stepBefore";
  return LineInterpolation2;
})(LineInterpolation || {});
var ScaleDistribution = /* @__PURE__ */ ((ScaleDistribution2) => {
  ScaleDistribution2["Linear"] = "linear";
  ScaleDistribution2["Log"] = "log";
  ScaleDistribution2["Ordinal"] = "ordinal";
  ScaleDistribution2["Symlog"] = "symlog";
  return ScaleDistribution2;
})(ScaleDistribution || {});
var GraphGradientMode = /* @__PURE__ */ ((GraphGradientMode2) => {
  GraphGradientMode2["Hue"] = "hue";
  GraphGradientMode2["None"] = "none";
  GraphGradientMode2["Opacity"] = "opacity";
  GraphGradientMode2["Scheme"] = "scheme";
  return GraphGradientMode2;
})(GraphGradientMode || {});
var StackingMode = /* @__PURE__ */ ((StackingMode2) => {
  StackingMode2["None"] = "none";
  StackingMode2["Normal"] = "normal";
  StackingMode2["Percent"] = "percent";
  return StackingMode2;
})(StackingMode || {});
var BarAlignment = /* @__PURE__ */ ((BarAlignment2) => {
  BarAlignment2[BarAlignment2["After"] = 1] = "After";
  BarAlignment2[BarAlignment2["Before"] = -1] = "Before";
  BarAlignment2[BarAlignment2["Center"] = 0] = "Center";
  return BarAlignment2;
})(BarAlignment || {});
var ScaleOrientation = /* @__PURE__ */ ((ScaleOrientation2) => {
  ScaleOrientation2[ScaleOrientation2["Horizontal"] = 0] = "Horizontal";
  ScaleOrientation2[ScaleOrientation2["Vertical"] = 1] = "Vertical";
  return ScaleOrientation2;
})(ScaleOrientation || {});
var ScaleDirection = /* @__PURE__ */ ((ScaleDirection2) => {
  ScaleDirection2[ScaleDirection2["Down"] = -1] = "Down";
  ScaleDirection2[ScaleDirection2["Left"] = -1] = "Left";
  ScaleDirection2[ScaleDirection2["Right"] = 1] = "Right";
  ScaleDirection2[ScaleDirection2["Up"] = 1] = "Up";
  return ScaleDirection2;
})(ScaleDirection || {});
const defaultLineStyle = {
  dash: []
};
var GraphTresholdsStyleMode = /* @__PURE__ */ ((GraphTresholdsStyleMode2) => {
  GraphTresholdsStyleMode2["Area"] = "area";
  GraphTresholdsStyleMode2["Dashed"] = "dashed";
  GraphTresholdsStyleMode2["DashedAndArea"] = "dashed+area";
  GraphTresholdsStyleMode2["Line"] = "line";
  GraphTresholdsStyleMode2["LineAndArea"] = "line+area";
  GraphTresholdsStyleMode2["Off"] = "off";
  GraphTresholdsStyleMode2["Series"] = "series";
  return GraphTresholdsStyleMode2;
})(GraphTresholdsStyleMode || {});
var LegendDisplayMode = /* @__PURE__ */ ((LegendDisplayMode2) => {
  LegendDisplayMode2["Hidden"] = "hidden";
  LegendDisplayMode2["List"] = "list";
  LegendDisplayMode2["Table"] = "table";
  return LegendDisplayMode2;
})(LegendDisplayMode || {});
const defaultReduceDataOptions = {
  calcs: []
};
var VizOrientation = /* @__PURE__ */ ((VizOrientation2) => {
  VizOrientation2["Auto"] = "auto";
  VizOrientation2["Horizontal"] = "horizontal";
  VizOrientation2["Vertical"] = "vertical";
  return VizOrientation2;
})(VizOrientation || {});
const defaultOptionsWithTimezones = {
  timezone: []
};
var BigValueColorMode = /* @__PURE__ */ ((BigValueColorMode2) => {
  BigValueColorMode2["Background"] = "background";
  BigValueColorMode2["None"] = "none";
  BigValueColorMode2["Value"] = "value";
  return BigValueColorMode2;
})(BigValueColorMode || {});
var BigValueGraphMode = /* @__PURE__ */ ((BigValueGraphMode2) => {
  BigValueGraphMode2["Area"] = "area";
  BigValueGraphMode2["Line"] = "line";
  BigValueGraphMode2["None"] = "none";
  return BigValueGraphMode2;
})(BigValueGraphMode || {});
var BigValueJustifyMode = /* @__PURE__ */ ((BigValueJustifyMode2) => {
  BigValueJustifyMode2["Auto"] = "auto";
  BigValueJustifyMode2["Center"] = "center";
  return BigValueJustifyMode2;
})(BigValueJustifyMode || {});
var BigValueTextMode = /* @__PURE__ */ ((BigValueTextMode2) => {
  BigValueTextMode2["Auto"] = "auto";
  BigValueTextMode2["Name"] = "name";
  BigValueTextMode2["None"] = "none";
  BigValueTextMode2["Value"] = "value";
  BigValueTextMode2["ValueAndName"] = "value_and_name";
  return BigValueTextMode2;
})(BigValueTextMode || {});
var TableCellDisplayMode = /* @__PURE__ */ ((TableCellDisplayMode2) => {
  TableCellDisplayMode2["Auto"] = "auto";
  TableCellDisplayMode2["BasicGauge"] = "basic";
  TableCellDisplayMode2["ColorBackground"] = "color-background";
  TableCellDisplayMode2["ColorBackgroundSolid"] = "color-background-solid";
  TableCellDisplayMode2["ColorText"] = "color-text";
  TableCellDisplayMode2["GradientGauge"] = "gradient-gauge";
  TableCellDisplayMode2["Image"] = "image";
  TableCellDisplayMode2["JSONView"] = "json-view";
  TableCellDisplayMode2["LcdGauge"] = "lcd-gauge";
  return TableCellDisplayMode2;
})(TableCellDisplayMode || {});
var TooltipDisplayMode = /* @__PURE__ */ ((TooltipDisplayMode2) => {
  TooltipDisplayMode2["Multi"] = "multi";
  TooltipDisplayMode2["None"] = "none";
  TooltipDisplayMode2["Single"] = "single";
  return TooltipDisplayMode2;
})(TooltipDisplayMode || {});
var SortOrder = /* @__PURE__ */ ((SortOrder2) => {
  SortOrder2["Ascending"] = "asc";
  SortOrder2["Descending"] = "desc";
  SortOrder2["None"] = "none";
  return SortOrder2;
})(SortOrder || {});
const defaultVizLegendOptions = {
  calcs: []
};
var BarGaugeDisplayMode = /* @__PURE__ */ ((BarGaugeDisplayMode2) => {
  BarGaugeDisplayMode2["Basic"] = "basic";
  BarGaugeDisplayMode2["Gradient"] = "gradient";
  BarGaugeDisplayMode2["Lcd"] = "lcd";
  return BarGaugeDisplayMode2;
})(BarGaugeDisplayMode || {});
const defaultTableFieldOptions = {
  align: "auto",
  displayMode: "auto" /* Auto */,
  inspect: false
};

const defaultAnnotationTarget = {
  tags: []
};
const defaultAnnotationQuery = {
  builtIn: 0,
  enable: true,
  hide: false,
  showIn: 0,
  type: "dashboard"
};
const defaultDashboardLink = {
  asDropdown: false,
  includeVars: false,
  keepTime: false,
  tags: [],
  targetBlank: false
};
const defaultGridPos = {
  h: 9,
  w: 12,
  x: 0,
  y: 0
};
const defaultThresholdsConfig = {
  steps: []
};
const defaultDashboardCursorSync = 0 /* Off */;
const defaultPanel$1 = {
  links: [],
  repeatDirection: "h",
  tags: [],
  targets: [],
  thresholds: [],
  timeRegions: [],
  transformations: [],
  transparent: false
};
const defaultFieldConfigSource$1 = {
  overrides: []
};
const defaultMatcherConfig = {
  id: ""
};
const defaultFieldConfig$1 = {
  links: [],
  mappings: []
};
const defaultRowPanel = {
  collapsed: false,
  panels: []
};
const defaultDashboard$1 = {
  editable: true,
  graphTooltip: 0 /* Off */,
  links: [],
  panels: [],
  schemaVersion: 36,
  style: "dark",
  tags: [],
  timezone: "browser"
};

const defaultDashboard = defaultDashboard$1;
const defaultPanel = defaultPanel$1;
const defaultFieldConfig = defaultFieldConfig$1;
const defaultFieldConfigSource = defaultFieldConfigSource$1;

const defaultPlaylist = {
  interval: "5m",
  items: []
};

exports.AxisColorMode = AxisColorMode;
exports.AxisPlacement = AxisPlacement;
exports.BarAlignment = BarAlignment;
exports.BarGaugeDisplayMode = BarGaugeDisplayMode;
exports.BigValueColorMode = BigValueColorMode;
exports.BigValueGraphMode = BigValueGraphMode;
exports.BigValueJustifyMode = BigValueJustifyMode;
exports.BigValueTextMode = BigValueTextMode;
exports.GraphDrawStyle = GraphDrawStyle;
exports.GraphGradientMode = GraphGradientMode;
exports.GraphTransform = GraphTransform;
exports.GraphTresholdsStyleMode = GraphTresholdsStyleMode;
exports.LegendDisplayMode = LegendDisplayMode;
exports.LineInterpolation = LineInterpolation;
exports.ScaleDirection = ScaleDirection;
exports.ScaleDistribution = ScaleDistribution;
exports.ScaleOrientation = ScaleOrientation;
exports.SortOrder = SortOrder;
exports.StackingMode = StackingMode;
exports.TableCellDisplayMode = TableCellDisplayMode;
exports.TooltipDisplayMode = TooltipDisplayMode;
exports.VisibilityMode = VisibilityMode;
exports.VizOrientation = VizOrientation;
exports.defaultAnnotationQuery = defaultAnnotationQuery;
exports.defaultAnnotationTarget = defaultAnnotationTarget;
exports.defaultDashboard = defaultDashboard;
exports.defaultDashboardCursorSync = defaultDashboardCursorSync;
exports.defaultDashboardLink = defaultDashboardLink;
exports.defaultFieldConfig = defaultFieldConfig;
exports.defaultFieldConfigSource = defaultFieldConfigSource;
exports.defaultGridPos = defaultGridPos;
exports.defaultLineStyle = defaultLineStyle;
exports.defaultMatcherConfig = defaultMatcherConfig;
exports.defaultOptionsWithTimezones = defaultOptionsWithTimezones;
exports.defaultPanel = defaultPanel;
exports.defaultPlaylist = defaultPlaylist;
exports.defaultReduceDataOptions = defaultReduceDataOptions;
exports.defaultRowPanel = defaultRowPanel;
exports.defaultTableFieldOptions = defaultTableFieldOptions;
exports.defaultThresholdsConfig = defaultThresholdsConfig;
exports.defaultVizLegendOptions = defaultVizLegendOptions;
