// @grafana/ui/dist/ReactMonacoEditor-999d05ce.js
'use strict';

var MonacoEditor = require('@monaco-editor/react');
var React = require('react');
var index = require('./index.js');
require('@grafana/data');
require('memoize-one');
require('@emotion/react');
require('tinycolor2');
require('@emotion/css');
require('classnames');
require('react-inlinesvg');
require('hoist-non-react-statics');
require('react-popper-tooltip');
require('react-dom');
require('lodash');
require('slate');
require('ansicolor');
require('react-select');
require('react-select/async');
require('react-select/creatable');
require('react-custom-scrollbars-2');
require('react-transition-group');
require('react-popper');
require('react-transition-group/Transition');
require('@grafana/e2e-selectors');
require('i18next');
require('react-i18next');
require('rc-cascader');
require('react-use/lib/useMeasure');
require('react-select/async-creatable');
require('react-window');
require('@react-aria/focus');
require('react-colorful');
require('react-use');
require('date-fns/formatDuration');
require('@react-aria/button');
require('@react-aria/menu');
require('@react-stately/menu');
require('@react-aria/dialog');
require('@react-aria/overlays');
require('rc-time-picker');
require('react-calendar');
require('slate-plain-serializer');
require('slate-react');
require('uuid');
require('is-hotkey');
require('prismjs');
require('immutable');
require('calculate-size');
require('react-highlight-words');
require('rxjs');
require('rxjs/operators');
require('react-table');
require('react-use/lib/usePrevious');
require('@grafana/schema');
require('uplot');
require('jquery');
require('@react-aria/utils');
require('react-use/lib/useAsync');
require('@sentry/browser');
require('react-use/lib/useClickAway');
require('rc-drawer');
require('rc-slider');
require('rc-tooltip');
require('react-hook-form');
require('react-router-dom');
require('react-dropzone');
require('ol/format/WKT');
require('ol/geom');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var MonacoEditor__default = /*#__PURE__*/_interopDefaultLegacy(MonacoEditor);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function getColors(theme) {
  if (theme === void 0) {
    return {};
  } else {
    return {
      "editor.background": theme.components.input.background,
      "minimap.background": theme.colors.background.secondary
    };
  }
}
function defineThemes(monaco, theme) {
  const colors = getColors(theme);
  monaco.editor.defineTheme("grafana-dark", {
    base: "vs-dark",
    inherit: true,
    colors,
    rules: [
      { token: "predefined", foreground: theme == null ? void 0 : theme.visualization.getColorByName("purple") },
      { token: "operator", foreground: theme == null ? void 0 : theme.visualization.getColorByName("orange") },
      { token: "tag", foreground: theme == null ? void 0 : theme.visualization.getColorByName("green") }
    ]
  });
  monaco.editor.defineTheme("grafana-light", {
    base: "vs",
    inherit: true,
    colors,
    rules: [
      { token: "predefined", foreground: theme == null ? void 0 : theme.visualization.getColorByName("purple") },
      { token: "operator", foreground: theme == null ? void 0 : theme.visualization.getColorByName("orange") },
      { token: "tag", foreground: theme == null ? void 0 : theme.visualization.getColorByName("green") }
    ]
  });
}

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
let initalized = false;
function initMonaco() {
  var _a;
  if (initalized) {
    return;
  }
  MonacoEditor.loader.config({
    paths: {
      vs: ((_a = window.__grafana_public_path__) != null ? _a : "public/") + "lib/monaco/min/vs"
    }
  });
  initalized = true;
  MonacoEditor.loader.init().then((monaco) => {
    defineThemes(monaco);
  });
}
const ReactMonacoEditor = (props) => {
  const theme = index.useTheme2();
  const monaco = MonacoEditor.useMonaco();
  React.useEffect(() => {
    if (monaco != null) {
      defineThemes(monaco, theme);
    }
  }, [monaco, theme]);
  initMonaco();
  const monacoTheme = theme.isDark ? "grafana-dark" : "grafana-light";
  return /* @__PURE__ */ React__default["default"].createElement(MonacoEditor__default["default"], __spreadValues({
    theme: monacoTheme
  }, props));
};

exports.ReactMonacoEditor = ReactMonacoEditor;
